### PR TITLE
Optimize RestrictionParser performance

### PR DESCRIPTION
**Pull request by SiarheiFedartsou**
# Issue

closes https://github.com/Project-OSRM/osrm-backend/issues/4102

## Tasklist

 - [ ] CHANGELOG.md entry ([How to write a changelog entry](http://keepachangelog.com/en/1.0.0/#how))
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [ ] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md))
 - [ ] review
 - [ ] adjust for comments
 - [ ] cherry pick to release branch

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/extractor/restriction_parser.cpp [hunk 1]
```diff
@@ -59,31 +75,13 @@ RestrictionParser::TryParse(const osmium::Relation &relation) const
...
-        filter.add(true, "restriction:" + namespaced);
-    }
-
     const osmium::TagList &tag_list = relation.tags();
 
     osmium::tags::KeyFilter::iterator fi_begin(filter, tag_list.begin(), tag_list.end());
     osmium::tags::KeyFilter::iterator fi_end(filter, tag_list.end(), tag_list.end());
 
     // if it's not a restriction, continue;
-    if (std::distance(fi_begin, fi_end) == 0)
```
Comment (SiarheiFedartsou) [inline src/extractor/restriction_parser.cpp:86]: It is O(N) for this iterator if I understand it correctly.

File: src/extractor/restriction_parser.cpp [hunk 2]
```diff
@@ -126,6 +126,25 @@ RestrictionParser::TryParse(const osmium::Relation &relation) const
         {
             return {};
```
Comment (SiarheiFedartsou) [inline src/extractor/restriction_parser.cpp:127]: Noticed that here we do `return {}` if found something unexpected in tag values, should we just ignore it instead? may be with some additional warning

File: src/extractor/restriction_parser.cpp [hunk 3]
```diff
@@ -126,6 +126,25 @@ RestrictionParser::TryParse(const osmium::Relation &relation) const
...
+
+            if (parsed.empty())
+                continue;
+
+            for (const auto &p : parsed)
+            {
+                std::vector<util::OpeningHours> hours = util::ParseOpeningHours(p.condition);
+                // found unrecognized condition, continue
+                if (hours.empty())
+                    return {};
```
Comment (SiarheiFedartsou) [inline src/extractor/restriction_parser.cpp:143]: The same, if we found something unexpected we just return, should it be just ignoring instead?

File: src/extractor/restriction_parser.cpp [hunk 4]
```diff
@@ -99,18 +97,20 @@ RestrictionParser::TryParse(const osmium::Relation &relation) const
     bool is_multi_from = false;
     bool is_multi_to = false;
 
+    std::vector<util::OpeningHours> condition;
+
     for (; fi_begin != fi_end; ++fi_begin)
     {
-        const std::string key(fi_begin->key());
```
Comment (SiarheiFedartsou) [inline src/extractor/restriction_parser.cpp:104]: Curious why didn’t we have warning about unused `key` variable here 🤔

File: include/extractor/restriction_parser.hpp [hunk 5]
```diff
@@ -43,15 +45,16 @@ class RestrictionParser
...
     std::vector<InputTurnRestriction> TryParse(const osmium::Relation &relation) const;
 
   private:
     bool ShouldIgnoreRestriction(const std::string &except_tag_string) const;
 
     bool use_turn_restrictions;
     bool parse_conditionals;
-    std::vector<std::string> restrictions;
+    std::set<std::string> restrictions;
+    osmium::tags::KeyFilter filter;
```
Comment (mjjbell) [inline include/extractor/restriction_parser.hpp:57]: The Osmium docs say this is [deprecated](https://docs.osmcode.org/libosmium/latest/namespaceosmium_1_1tags.html#a75df2aa39d225085eeffd438b701bfa1), and should be replaced with `osmium::TagsFilter`.

File: src/extractor/restriction_parser.cpp (hunk 2 above)
Comment (mjjbell) [reply src/extractor/restriction_parser.cpp:127]: In most cases you'll get at most one restriction returned per relation, it's only no entry/exits that could have multiple.

I think the way to view it is, if the relation is invalid, how correct will the remaining parsed restrictions be? It's difficult to know, so it's cleaner to not to try and deal with them.

We could add a log, but I would only make it a debug level. It's not OSRM's problem if input data is not correct.

**Comment by mjjbell** [comment]
Can we run a benchmark on a reasonably sized OSM extract to evaluate the performance?

**Comment by SiarheiFedartsou** [comment]
> Can we run a benchmark on a reasonably sized OSM extract to evaluate the performance?

Well, I ran it on 1.5GB poland-latest.osm.pbf(let me know if I should use something larger). I ran it for 3 times on both current master and this PR and used this value as a metric(because related code is executed only before this log):
Screenshot 2022-08-29 at 18 18 26



On current master(on my M1 Pro + 32GB + 10 threads with `./osrm-extract -p ../profiles/car.lua poland-latest.osm.pbf`):
```
(35.713 + 36.4149 +37.1714) / 3 = 36.4331
```

With this PR:
```
(36.2307 + 35.4937 + 36.8094) / 3 = 36.1780
```

Results are definitely not impressive.

File: src/extractor/restriction_parser.cpp [hunk 6]
```diff
@@ -270,17 +264,25 @@ bool RestrictionParser::ShouldIgnoreRestriction(const std::string &except_tag_st
...
-        except_tag_string.begin(), except_tag_string.end(), delimiter_re, -1);
-    std::sregex_token_iterator except_tags_end;
-
-    return std::any_of(except_tags_begin, except_tags_end, [&](const std::string &current_string) {
-        return std::end(restrictions) !=
-               std::find(std::begin(restrictions), std::end(restrictions), current_string);
-    });
+    // split `except_tag_string` by semicolon and check if any of items is in `restrictions`
+    std::string current_string;
+    for (size_t index = 0; index < except_tag_string.size(); index++)
```
Comment (mjjbell) [inline src/extractor/restriction_parser.cpp:269]:
```suggestion
    for (auto index : util::irange<size_t>(0, except_tag_string.size())
```

**Review by mjjbell — APPROVED** [review]

**Comment by mjjbell** [comment]
> Results are definitely not impressive.

It's to be expected given the number of restrictions is small relative to other elements being parsed. In any case, it's tidying up the code for future improvements :+1:

**Comment by SiarheiFedartsou** [comment]
@mjjbell it seems this new TagsParser is incompatible with clang 6. I'd propose to go back to KeysFilter(as I understand bumping the compiler is not an option). WDYT?
Screenshot 2022-08-29 at 23 33 36

**Comment by mjjbell** [comment]
> I'd propose to go back to KeysFilter(as I understand bumping the compiler is not an option). WDYT?

:+1:

**Pull request merged**